Alternate approach for HELPER_METHOD_FRAME removal for arithmetic div…

davidwrighton opened the pull request:
This is a combination of @am11's work in PR #109087 and some work to just rewrite the Windows X86 helpers in assembly.
  - For non 64-bit platforms, remove the helpers
  - For Unix platforms, rely on an implementation which uses an FCall to native code to invoke the various operations.
  - For Windows X86 on CoreCLR, rewrite the helpers in assembly with a tail-call approach for throwing.
    - Also it was noted that the existing helpers do a fair bit of unnecessary stack manipulation, and the manual rewrite avoids all of that. Maybe this will actually be faster. (Turns out it is about the same on performance. The real win would come from re-ordering the argument order, but that's a much more invasive change in the JIT, which I judge to be too risky)

davidwrighton committed:
Alternate approach for HELPER_METHOD_FRAME removal for arithmetic divide operations
- Instead of moving to managed
  - For non 64-bit platforms, remove the helpers
  - For Unix platforms, rely on [[clang::musttail]] to tail-call a helper which throws (current implementation still uses FCThrow, but that should be relatively easy to move to a call to the managed throw routines
  - For Windows X86, rewrite the helpers in assembly with a tail-call approach for throwing.
    - As part of writing the helpers in assembly, it was noted that the stack layout is swapped from the underyling CRT routines for 64bit arithmetic. Tweak the helper calls for those cases to pass in data in the order which the CRT is going to handle them efficiently. Do this for all architectures/OSs as other ones don't really care about this weirdness.
    - Also it was noted that the existing helpers do a fair bit of unnecessary stack manipulation, and the manual rewrite avoids all of that. Maybe this will actually be faster.

TODO: Bump jit version
TODO: Bump R2R version
TODO: Actually use managed helper to throw instead of keeping the FCThrow in place

davidwrighton committed:
Update alignment to be a bit better for the assembly routines

davidwrighton committed:
Try to adjust jithelpers.cpp for better building on arm32

jkoritzinsky commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -3642,3 +3466,250 @@ void InitJITHelpers2()
+
+#if defined(TARGET_32BIT) && !(defined(TARGET_X86) && defined(TARGET_WINDOWS))
+
+// Using the musttail attribute requires that we disble our protection against using return in the wrong place
+#ifdef return
+#undef return
+#endif
```
Related: I'm trying to remove the cases outside of HMFs that engage this infra as it continues to cause us problems in places.

am11 commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -3642,3 +3466,250 @@ void InitJITHelpers2()
+    {
+        if (divisor == 0)
+        {
+            [[clang::musttail]] return JIT_ThrowDivideByZero_RetInt();
```
We can define MUSTTAIL for this:
```c++
#if TARGET_UNIX
#ifdef __clang__
#define MUSTTAIL [[clang::musttail]]
#else // __clang__
#define MUSTTAIL [[gnu::musttail]]
#endif // __clang__
#endif // TARGET_UNIX
```

am11 commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -3642,3 +3466,250 @@ void InitJITHelpers2()
+{
+    FCALL_CONTRACT;
+
+    FCThrow(kOverflowException);
```
Could we directly inline them in JIT? there already are `CORINFO_HELP_OVERFLOW` and `CORINFO_HELP_THROWDIVZERO` exposed.

davidwrighton replied:
Yes. This is just an experiment to see how musttail works, and if it can be used to help solve the HMF problem without changing the JIT/etc for this issue. Given the number of test failures, something clearly didn't go completely right with my change here, so I want to understand what I've missed.

davidwrighton committed:
Fix more oops in the build

davidwrighton committed:
More attempts to fix build breaks... maybe I should actually set up an arm32 build environment...

davidwrighton committed:
Get the types right for JIT_UDiv and JIT_UMod

jkotas commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -34,19 +34,24 @@
 
     JITHELPER(CORINFO_HELP_UNDEF,               NULL,               METHOD__NIL)
 
+    // CORINFO_HELP_DBL2INT, CORINFO_HELP_DBL2UINT, and CORINFO_HELP_DBL2LONG get
+    // patched for CPUs that support SSE2 (P4 and above).
```
```suggestion
```
Outdated comment

davidwrighton committed:
Swap the args around for the 64bit long/ulong divide helpers back to the original order (Sacrifices perf, but avoid needinga  bunch of optimizer changes)
Adjust the negate path for the X86 hand coded assembly for JIT_LDiv to load edx before jumping to JIT_LDiv_DoNegate

davidwrighton committed:
Attempt to fix 32bit linux builds, and remove all use of FCThrow, and the FCThrow machinery

davidwrighton committed:
Fix BCL build

davidwrighton committed:
Merge branch 'main' into alternate_approach_to_removingHMF_for_arithmetic

davidwrighton committed:
Get this approach building on Linux X86 and Linux Arm ... turns out that the musttail feature doesn't work on Linux X86 very well, but it DOES seem to work on arm, so there is at least 1 platform where it works

davidwrighton committed:
Merge branch 'main' into alternate_approach_to_removingHMF_for_arithmetic

am11 commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -3618,3 +3426,212 @@ void InitJITHelpers2()
+
+#if TARGET_UNIX
+#ifdef __clang__
+#define MUSTTAIL [[clang::musttail]]
```
BTW, the performance concern was only for win-x86 and performance optimizations aren't critical for linux-arm and linux-x86. Moreover, it seems like MUSTTAIL is only to accelerate exception path with no effect on the actual div/mod ops? If this is the case, then I think we can just drop this optimization even if it was for win-x86 because perf concern was about the div/mod (non-exception) happy path.

davidwrighton replied:
The MUSTTAIL logic here isn't an optimization as without HMFs we can't through an exception from an FCALL, but I agree, it doesn't look like it has worked out as a reasonable way to make this work simple. My current plan is to keep the work I've done to port the Windows X86 helpers to handcoded assembly, and merge it with @am11's work that is more general purpose, and then call it good.

davidwrighton committed:
Integrate am11's work into my branch

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into alternate_approach_to_removingHMF_for_arithmetic

am11 commented:
There was also a pure managed experiment somewhere in the middle using generic math with no FCall that was passing all tests: https://github.com/dotnet/runtime/blob/e51389e0973e68f7033b7947a8b689a928e55956/src/libraries/System.Private.CoreLib/src/System/Math.DivModInt.cs#L180-L243 (those InternalCalls after line 243 are unused)

It was quite slow https://github.com/EgorBot/runtime-utils/issues/193. (was probably dreaming that it could outperform the native implementation 🥲)

The FCall version you've applied was performing much better https://github.com/EgorBot/runtime-utils/issues/191, which is fine for 32-bit linux.

am11 commented on src/libraries/System.Private.CoreLib/src/System/Math.DivModInt.cs:
```diff
@@ -0,0 +1,224 @@
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern uint DivUInt32Internal(uint dividend, uint divisor);
+
+#if NATIVEAOT
```
AOT could use InternalCall (with revert of `RuntimeHelpers.cs` file). There was a slight regression with QCall on coreclr.

jkotas replied:
Right, my point is that there is no reason for two to be different.

(I do not have a strong opinion on FCall vs. QCall here as long as it is the same.)

davidwrighton committed:
Fix x86 eh stuff

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into alternate_approach_to_removingHMF_for_arithmetic

davidwrighton committed:
Move the math helpers to be consistently FCalls

jkotas commented on src/coreclr/nativeaot/Runtime/MathHelpers.cpp:
```diff
@@ -56,29 +56,33 @@ FCIMPL1_D(uint32_t, RhpDbl2UInt, double val)
 
 #ifndef HOST_64BIT
-EXTERN_C int64_t QCALLTYPE RhpLDiv(int64_t i, int64_t j)
+FCIMPL2(int64_t, DivInt64Internal, int64_t i, int64_t j)
```
Does this need `_VV` flavor of the macro to get the arguments ordered correctly on all arches?

davidwrighton replied:
Agh, I thought I had done that.

davidwrighton committed:
Tweak macros

am11 commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/CompilerServices/RuntimeHelpers.cs:
```diff
@@ -18,5 +18,7 @@ public static int OffsetToStringData
 
         [Intrinsic]
         public static extern void InitializeArray(Array array, RuntimeFieldHandle fldHandle);
+
+        public const string QCall = "*";
```
```suggestion
```

(no longer needed)

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into alternate_approach_to_removingHMF_for_arithmetic

davidwrighton committed:
Code revewi from am11

am11 commented:
@EgorBot -windows_intel -use32bit

```c#
using BenchmarkDotNet.Attributes;

public class Bench
{
    long a = 10000000000000000L;
    long b = 1000L;

    [Benchmark]
    public long SDiv() => a / b;
}
```

am11 reviewed (approved):
Number are looking promising https://github.com/EgorBot/runtime-utils/issues/336#issuecomment-2770739898 (0.55) 🚀

am11 commented on src/coreclr/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:
```diff
@@ -1,6 +1,7 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
```
nit: undo

davidwrighton committed:
Remove unused System namespace import

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into alternate_approach_to_removingHMF_for_arithmetic

jkotas commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -55,14 +65,25 @@
+#else // TARGET_64BIT
+    DYNAMICJITHELPER(CORINFO_HELP_LMUL_OVF,     NULL,               METHOD__NIL)
+    DYNAMICJITHELPER(CORINFO_HELP_ULMUL_OVF,    NULL,               METHOD__NIL)
+    JITHELPER(CORINFO_HELP_LDIV,                NULL,           METHOD__NIL)
```
Padding for the `METHOD__NIL` column is off.

jkotas commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -55,14 +65,25 @@
+    DYNAMICJITHELPER(CORINFO_HELP_ULMOD,        NULL,               METHOD__MATH__MOD_UINT64)
+#endif // TARGET_X86 && TARGET_WINDOWS
+#else // TARGET_64BIT
+    DYNAMICJITHELPER(CORINFO_HELP_LMUL_OVF,     NULL,               METHOD__NIL)
```
If `CORINFO_HELP_LMUL/ULMUL_OVF` are unused on 64-bit, I would expect them to be `JITHELPER`.

If they are used, I would expect them to look like the 32-bit ones and be outside the `#ifndef TARGET_64BIT` ifdef.

davidwrighton committed:
AAGGHH

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into alternate_approach_to_removingHMF_for_arithmetic

davidwrighton committed:
Code review comments

jkotas reviewed (approved):
Thanks!

am11 commented:
Thanks for this @davidwrighton!

For the remaining alloc-like HMFs, I think we can either use managed `RuntimeTypeHandle_InternalAlloc{NoChecks}` or the assembly ones in `AllocFast.{asm,S}` (+ the FOH):
https://github.com/dotnet/runtime/blob/b3188c58e818235d68ededa65abf64a03b5a7d21/src/coreclr/vm/jithelpers.cpp#L622
cc @AaronRobinsonMSFT

akoeplinger commented:
@davidwrighton @am11 this causes an issue on the runtime sdk codeflow on win-x86: https://github.com/dotnet/sdk/pull/48034

```
finalizer.obj : error LNK2001: unresolved external symbol RhpDbl2Int [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol RhpDbl2Lng [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol RhpLng2Dbl [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol RhpULng2Dbl [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol ceil [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol cos [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol floor [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol modf [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol pow [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol sin [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
finalizer.obj : error LNK2001: unresolved external symbol tan [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
D:\a\_work\1\vmr\src\sdk\artifacts\bin\finalizer\Release\net10.0-windows\native\finalizer.exe : fatal error LNK1120: 11 unresolved externals [D:\a\_work\1\vmr\src\sdk\src\Layout\finalizer\finalizer.csproj]
```

am11 commented:
Seems like it is using old ILCompiler host (win-x64) package for `ilc.exe` with newer runtime (win-x86) packages for aotsdk/framework? That kind of mixing won't work when we change any internal export. If that is the problem then fix is to update the host package to the same version as runtime packages. i.e. `runtime.win-x86.Microsoft.DotNet.ILCompiler` should match win-x64 host's `Microsoft.DotNet.ILCompiler` version. We cannot mix and match; it's not everyday these exports change but they can change any day, so it'd be a flawed assumption in source-build design to use mismatched versions.

am11 commented:
Ah, nvm, it is also failing in AOT outerloop: https://github.com/dotnet/runtime/runs/40052903604. Although we haven't touched these exports. 👀

akoeplinger commented:
Ok thanks, I'll prepare a revert since this is blocking the codeflow.
